File: HW5/Q3.m
clear all;
clc;

rho = [0:9] / 10; 
%rho = [rho 0.91 0.92 0.93 0.94 0.95];
Nr = 4;
Nt = 4;

samples = 1000;

H = zeros(Nr, Nt);

avg_condition_independent = [];
avg_condition_correlated = [];

for n = 1 : length(rho)
    avg = 0;
    for sample = 1 : samples
        for i = 1 : Nr
            for j = 1 : Nt
                H(i, j) = rand + 1j*rand;
            end
        end
        singular_values = svd(H);
        avg = avg + max(singular_values) / min(singular_values);
    end
    avg_condition_independent = [avg_condition_independent avg / samples];
end


for n = 1 : length(rho)
    avg = 0;
    R = [1        rho(n)    rho(n)^2 rho(n)^3;
         rho(n)   1         rho(n)   rho(n)^2;
         rho(n)^2 rho(n)    1        rho(n)  ;
         rho(n)^3 rho(n)^2  rho(n)   1       ];

    for sample = 1 : samples
        for i = 1 : Nr
            for j = 1 : Nt
                H(i, j) = rand + 1j*rand;
            end
        end
        singular_values = svd(sqrtm(R) * H * sqrtm(R));
        avg = avg + max(singular_values) / min(singular_values);
    end
    avg_condition_correlated = [avg_condition_correlated avg / samples];
end

figure
hold on
grid on
plot(rho, avg_condition_independent);
plot(rho, avg_condition_correlated);
xlabel("\rho");
ylabel("condition number")
legend("i.i.d rayleigh fading", "correlated");

File: HW5/Q3.py
import numpy as np
from scipy.linalg import sqrtm
import matplotlib.pyplot as plt

# Parameters
rho = np.concatenate([np.arange(0, 1, 0.1)])  # rho values
Nr = 4
Nt = 4
samples = 1000

# Initialize results
avg_condition_independent = []
avg_condition_correlated = []

# Independent Rayleigh fading
for n in range(len(rho)):
    avg = 0
    for sample in range(samples):
        # Generate random H (i.i.d Rayleigh fading)
        H = np.sqrt(0.5) * (np.random.rand(Nr, Nt) + 1j * np.random.rand(Nr, Nt))
        singular_values = np.linalg.svd(H, compute_uv=False)  # Get singular values
        avg += max(singular_values) / min(singular_values)
    avg_condition_independent.append(avg / samples)

# Correlated Rayleigh fading
for n in range(len(rho)):
    avg = 0
    # Correlation matrix R
    R = np.array([
        [1, rho[n], rho[n]**2, rho[n]**3],
        [rho[n], 1, rho[n], rho[n]**2],
        [rho[n]**2, rho[n], 1, rho[n]],
        [rho[n]**3, rho[n]**2, rho[n], 1]
    ])
    R_sqrt = sqrtm(R)
    
    for sample in range(samples):
        # Generate random H
        H = np.sqrt(0.5) * (np.random.rand(Nr, Nt) + 1j * np.random.rand(Nr, Nt))
        
        correlated_H = np.dot(R_sqrt, np.dot(H, R_sqrt))  # Apply correlation
        singular_values = np.linalg.svd(correlated_H, compute_uv=False)
        avg += max(singular_values) / min(singular_values)
    avg_condition_correlated.append(avg / samples)

# Plot results
plt.figure()
plt.plot(rho, avg_condition_independent, label="i.i.d Rayleigh fading")
plt.plot(rho, avg_condition_correlated, label="Correlated Rayleigh fading")
plt.xlabel(r"$\rho$")
plt.ylabel("Condition number")
plt.legend()
plt.grid()
plt.show()
